### ISSUE TITLE
RC 0.10 3X Slower than 0.9 and Error Compiling From Source Under Certain Conditions

### ISSUE THREAD

**Issue by rob-rowe** [tensorflow/tensorflow#3603]
This issue is a follow-up to [this posting](http://stackoverflow.com/questions/38704095/tensorflow-rc-0-10-3x-slower-than-0-9) on stack overflow. As noted, the 0.10 version is ~3x slower than version 0.9 for a particular training script I am using. The same slow-down is observed when using the pip version with CuDNN4 or with a version compiled from source and CuDNN 5.1.

I'm not sure what information about the training script would help. It is a fairly complex script that is training a network for object detection using methods similar to YOLO and SSD. It won't be possible to post the entire set of code, but if there is a way to isolate the source of the delay I may be able to modify it and post that portion.

While trying to get insight into the source of the slow-down I also noted that I wasn't able to compile from source using CuDNN4 as noted in the SO posting. This is a secondary issue, but may also warrant investigation.

**Comment by rryan** [comment]
Could you produce a timeline and share that?  

I think this might be the current best how-to on how to produce timelines :) if anyone knows of a better one please share that:
https://github.com/tensorflow/tensorflow/issues/1824#issuecomment-225754659

I think that it's going to be hard to help without seeing your code. Would you be able to post your graph protobuf?

**Comment by rob-rowe** [comment]
@rryan I've attached the timelines named appropriately. Each was generated with the identical code running under either tf0.9 or tf0.10rc. The elapsed times for 100 training steps were 4:20 and 14:36, respectively. 
[timelines.tar.gz](https://github.com/tensorflow/tensorflow/files/397146/timelines.tar.gz)
For some reason, the timing routine gave repeated error messages when running under 0.9 but not for 0.10. The 0.9 message was: 

> W tensorflow/core/common_runtime/gpu/gpu_tracer.cc:513] Unhandled API Callback for 2 41

I've not used the timing functionality before. Let me know if I've done something incorrectly.

Also, what's the best way to generate a graph protobuf?

Thanks

**Comment by rob-rowe** [comment]
@rryan: Not sure if it matters but I've attached two other timelines taken in loops in which summary operations are not being performed
[timelines2.tar.gz](https://github.com/tensorflow/tensorflow/files/397509/timelines2.tar.gz)

**Comment by rryan** [comment]
@rob-rowe -- thanks! The easiest way to get a graph proto is probably to insert some code into your training script to write the graph to a file.

Once you're done building your graph, right where you would start your session (maybe with a supervisor, etc.) -- you could do this:

```python
# If you don't already have a handle to your graph:
g = tf.get_default_graph() 

with open('/tmp/graph.pbtxt', 'w') as f:
  f.write(str(g.as_graph_def()))
```

**Comment by rryan** [comment]
One thing that sticks out from the traces are these Sum ops associated with your batch norms:

* net_cnn/conv1/cbn/moments/moments/mean_ss
* net_cnn/conv1/cbn/moments/moments/var_ss
* net_cnn/conv2/cbn/moments/moments/mean_ss
* net_cnn/conv2/cbn/moments/moments/var_ss

They take a handful of millis in 0.9 but up to 700ms in 0.10. I would potentially chalk that up to a tracing noise (e.g. your Sum ops could waiting on some dependency that isn't ready yet -- hard to know w/o seeing the graph) but in your second trace it looks like every Sum op is taking much longer than normal. 

@prb12 I'm not super used to interpreting Chrome tracing timelines -- any gotchas to be aware of when interpreting this?

@rmlarsen Are you aware of any change to Eigen between 0.9 and 0.10 that could cause Sum (@rob-rowe -- is this float32? float16?) to do this with a Tesla K40c (mentioned in the Stack Overflow post)?

**Comment by rob-rowe** [comment]
@rryan --
graph proto attached. 
[graph.pbtxt.tar.gz](https://github.com/tensorflow/tensorflow/files/397621/graph.pbtxt.tar.gz)
The batch norm variables you mention are all default data types (float32)

**Comment by prb12** [comment]
@rryan Nop - these traces make it pretty clear that there is a perf regression in the Sum reduction kernel on GPU.  I think it would be useful to compare what CUDA launch parameters are being used for the reduction.  The quickest way to do that is to run the program under nvprof as follows:

`nvprof --print-gpu-trace program args....`  
or 
`nvprof -o /tmp/nvprof.out program args...`  to produce a trace which can be loaded into `nvvp`

I think that the Eigen `FullReductionKernel` was probably changed btw 0.9 and 0.10.

**Comment by prb12** [comment]
@bsteiner may know more about this.

**Comment by rob-rowe** [comment]
@rryan, @prb12  -- I've attached nvprof.out files for each of the two cases. I ran the profiling for just 5 batches to keep the files from getting too large but they're still too big to upload here. Please let me know if [this link](https://1drv.ms/u/s!AnX4PArwPRXWg9dmWvZieqWqQqCQbw) doesn't work

**Comment by prb12** [comment]
Unfortunately both those files seem to be corrupted - it looks like the process didn't exit cleanly.  Did you Ctrl-C ?  (I think that to get a clean nvprof trace you need to exit gracefully)

**Comment by rob-rowe** [comment]
Odd, I did let the process end on it's own. I used your nvprof - o method and waited until the cursor returned, which was a multisecond delay after the training finished which I assume was file IO. The raw files here seem to be able to be consumed by nvprof -i I think (see attached). Perhaps compression failed? I uploaded the files uncompressed to google drive at [this link](https://drive.google.com/open?id=0BwIAwD4S8krbZ1VDU0NtOVRGMjg). Please let me know if this doesn't work and if so perhaps an alternative method. Thanks
[nvprof_out_tf9.txt](https://github.com/tensorflow/tensorflow/files/397882/nvprof_out_tf9.txt)
[nvprof_out_tf10.txt](https://github.com/tensorflow/tensorflow/files/397883/nvprof_out_tf10.txt)

**Comment by rob-rowe** [comment]
I've also just rerun both cases and confirmed they seem to be ending cleanly: 

> ==28091==Generated result file: /tmp/nvprof_tf0.10.0rc0_v2.out

Both of these 'v2' files are in the same google drive folder as previous

**Comment by rob-rowe** [comment]
Sorry -- the tf0.10.0rc0_v2 file is wrong (I didn't switch TF versions) I'll re-upload momentarily

**Comment by rob-rowe** [comment]
nvprof_tf0.10.0rc0_v2.out is now correct

**Comment by untom** [comment]
Just chiming in to say that I also experience heavy slowdowns using TF 0.10 RC0. In My case, the slowdown is even more severe: ~10x when using a ResNet-like architecture, on Python 3.5 using cuDNN 4 and CUDA 7.5.

**Comment by pawelsubko** [comment]
Same here. 10X slowdown

**Comment by olesalscheider** [comment]
I can confirm that the SumOp kernel is much slower in 0.10.0 than it used to be in 0.9.0 on an Nvidia Titan X.

**Comment by prb12** [comment]
@rob-rowe Thanks for the nvprof output - it looks like there are two common shapes of Sum reduction ops in your trace and one of them is being evaluated very differently by Eigen in 0.10 resulting in much less GPU parallelism and about a 60x slowdown!  (3.3ms -> 178ms)

It would help us to make a quick repro if we knew the shapes of the operands to these ops...  
Do you happen to have this information to hand?  If not, we could either get if from the GraphDef or the StepStats proto.

You can save a graphdef with shape annotations using code like this:
`tf.train.write_graph(g.as_graph_def(add_shapes=True),  '/tmp', 'graphdef.pbtxt')`

(Since you have timelines, you already have the code to capture the stepstats proto - this can be written to a .pbtxt file.  It contains the runtime shapes of all of the Tensors.)

**Comment by rob-rowe** [comment]
@prb12 the graphdef with shape annotations is attached
[graph_with_shapes.pbtxt.tar.gz](https://github.com/tensorflow/tensorflow/files/400073/graph_with_shapes.pbtxt.tar.gz)

**Comment by rob-rowe** [comment]
@prb12 here is the runtime metadata jic that is helpful

[metadata.pbtxt.tar.gz](https://github.com/tensorflow/tensorflow/files/400094/metadata.pbtxt.tar.gz)

**Comment by benoitsteiner** [comment]
I think I found the source of the problem. A fix is on the way. I'll update this thread once the fix is merged.

**Comment by prb12** [comment]
@rob-rowe Thanks for the diagnostics.

For future reference, it looks like the first example of the slow Sum is this one:
```
node {
  name: "net_cnn/conv1/cbn/moments/moments/mean_ss"
  op: "Sum"
  input: "net_cnn/conv1/Conv2D"
  input: "net_cnn/conv1/cbn/moments/moments/mean_ss/reduction_indices"
```
Where reduction indices are [0,1,2] and the first input has shape [100, 127, 127, 30]

**Comment by rob-rowe** [comment]
@prb12 A convolutional batch norm factor in the first layer --- yeah, I can see why that would cause a major slow down if the op is slow! I'm looking forward to the fix. Thanks

**Comment by vrv** [comment]
@benoitsteiner any updates on this?

**Comment by benoitsteiner** [comment]
I have just started one last round of testing. If everything comes back clean this time I'll be able to submit the fix first tomorrow morning.

**Comment by benoitsteiner** [comment]
The performance regression should have been fixed by https://github.com/tensorflow/tensorflow/commit/0e12a0b0dbcf17da880402bc01d8511aa4568cda. Please reopen if you still find that RC 0.10 is slower than 0.9 on your GPU.

**Comment by passerbydj** [comment]
@benoitsteiner  is the fix be picked into branch r0.10?

**Comment by rafaljozefowicz** [comment]
https://github.com/tensorflow/tensorflow/blob/r0.10/tensorflow/core/kernels/cwise_ops.h the fix doesn't seem to be in 0.10 branch.

**Comment by vrv** [comment]
It should be now.
